perf(linter): skip rules that do not have any relevant node types

camchenry opened the pull request:
- part of https://github.com/oxc-project/oxc/issues/12223
- supersedes stack: https://github.com/oxc-project/oxc/pull/12228

This PR adds the ability to add node type information to indicate what node types a lint rule acts on. The lint rule runner can use this information to optimize running lint rules by skipping nodes, files, or rules that don't apply.

This is a simple version that is focused on creating the framework for automatically generating the rule runner code. This is not the end of optimizations. Currently only 102 rules have node type information, while 471 rules have no node type information and will not be skipped. Adding node type information for more rules will improve performance. We may also be able to skip more nodes and files by being clever about how we use this information. For example, filtering out any rules which do not run on any node type, if the file contains no relevant node types for the rule.

- Adds a new `RuleRunner` trait which includes information needed for indicating whether a rule has node type info or not
- Adds a `oxc_linter_codegen` task which is used to generate the implementation of `RuleRunner` trait for all lint rules. This is mostly written by AI, so worth a bit of skepticism, but I've tried to manually verify the results and it looks good (just lacking more complex detection).
  - Currently, only rules that contain a single top-level `if` statement (or chain of `else if`), or a single `match` statement will have their node type info generated. This should give us a high level of confidence, since these patterns are relatively easy to verify and identify, but it's only a subset of all rules. Next steps will be to add support for `let...else` 
- Adjusts the main linting loop to use the node type information to skip rules for node types that don't apply
- Updates CI to include a check for any uncommitted linter codegen changes

---

## Benchmarks

In the latest version of this PR, CodSpeed shows an up to +13% increase in linter performance across large and small files. In the real-world, I've seen a ~5-15% improvement in performance, but in some cases close to zero for very small files. In either case, there appears to be little downside and a lot of potential upside here.

<img width="769" height="242" alt="Screenshot 2025-08-24 at 9 29 30 PM" src="https://github.com/user-attachments/assets/d2c430c4-17a9-4606-85d5-c66c734975d9" />
<img width="757" height="308" alt="Screenshot 2025-08-24 at 9 30 24 PM" src="https://github.com/user-attachments/assets/5508464f-de55-4f4e-9ddd-58f215c4ce85" />
<img width="976" height="368" alt="Screenshot 2025-08-24 at 9 34 52 PM" src="https://github.com/user-attachments/assets/203f83c7-46f3-49f1-a1a8-e18f7ec8483b" />

overlookmotel reviewed (commented):
Superb!

Probably you intend to do this anyway, but I think need to use `syn` to parse the rule files. The regex may be, for example, missing the `else` branch in patterns like this:

```rs
if let AstKind::Function(func) = node.kind() {
    // ...
} else if let AstKind::ArrowFunctionExpression(arrow) = node.kind() {
    // ...
}
```

If it's currently missing out some types, that may slow things down a little.

On the other hand, some patterns aren't currently being caught by the regex, like this one:

```rs
let Some(ss) = node.kind().as_switch_statement() else { return };
```

(from [no_duplicate_case rule](https://github.com/oxc-project/oxc/blob/4f8506e0611000d5d27af2ae09c476d8fdc34695/crates/oxc_linter/src/rules/eslint/no_duplicate_case.rs#L83-L94))

Matching those would take more rules out of the "any type" category, and speed things up even more!

---

I've made a few small comments about the `AstTypesIter` iterator, but that's just fiddling at the edges.

---

Edit 7/9/25: This PR has been scaled back a lot from the original. The first version showed ~90% perf improvement. For anyone wondering, that's the reason for all our "amazing perf!" comments.

camchenry commented on crates/oxc_semantic/src/ast_types_bitset.rs:
```diff
@@ -66,6 +68,55 @@ impl AstTypesBitset {
+        self.current_word &= self.current_word - 1;
+        if global_bit_index > AST_TYPE_MAX as usize {
+            return self.next();
+        }
```
It shouldn't be. I'm going to remove this.

camchenry commented:
Thanks @overlookmotel for the early feedback. I've shifted this PR to be a simple initial version of the optimization now that we've proven this can actually have double digit performance impact. This will likely be the biggest PR since it adds the framework (CI, codegen, etc.) for doing this optimization. But I'm hoping that we can break the rest of this work into smaller PRs so we can better quantify the impact of each change.

overlookmotel reviewed (commented):
I've gone through this PR in some depth, as obviously it's critical that rules don't get skipped when they shouldn't be. I think I've found a couple of potential bugs.

The rest of my comments are less important, but I feel that any simplifications that can be made would be a benefit as they'd make the code easier to understand, so any logic flaws stick out more.

(personally I find `syn` a bit of a nightmare - it's complicated and verbose, which tends to obscure meaning)

camchenry commented on tasks/linter_codegen/src/main.rs:
```diff
@@ -0,0 +1,438 @@
+    let mut impls_buf = String::new();
+
+    // Build a set of valid AstType variant names to validate any textual fallbacks.
+    let _allowed_asttype_variants = get_asttype_variant_names().unwrap_or_default();
```
Yeah thinking about this more, I realize there isn't much point to validating this here. I think using Strings instead is simpler, because the Rust compiler will check the validity anyway.

camchenry commented on tasks/linter_codegen/src/main.rs:
```diff
@@ -0,0 +1,438 @@
+    let mapping = build_asttype_name_to_index_map(&ast_file);
+    if mapping.is_empty() {
+        return None;
+    }
```
Good catch! I ended up removing all of this code anyway since we can get away with just using the string name and relying on the Rust compiler to validate that all of the source code generated is correct.

overlookmotel commented:
I'm part-way through reviewing this. I'll finish review tomorrow.

overlookmotel reviewed (approved):
Looks good!

This more conservative approach of only matching the simplest `if` / `else` statements obviously reduces the perf impact a lot, but I do think it's a good approach to start minimal and build on it step-by-step. It'll be easier to verify correctness as add support for more/complex patterns in later PRs.

None of my comments below are blockers. Just some thoughts for future, and I have submitted follow-on PRs for some of them.

overlookmotel commented on tasks/linter_codegen/src/main.rs:
```diff
@@ -0,0 +1,330 @@
+        }
+        let stmt = &block.stmts[0];
+        let Stmt::Expr(Expr::If(ifexpr), _) = stmt else { return None };
+        let mut variants = BTreeSet::new();
+        let result = collect_if_chain_variants(ifexpr, &mut variants);
+        if result == CollectionResult::Incomplete || variants.is_empty() {
+            return None;
+        }
```
The pattern of passing `out: &mut BTreeSet<String>` through various free functions is not incorrect, but it's perhaps a little un-idiomatic. I wonder if it'd be preferable to make some of these functions methods of the `IfElseKindDetector` trait instead?

Or, if you want to share some of the logic with other detector structs in future PRs, add a `Variants` type which wraps a `BTreeSet<String>` and has methods like `extract_variants_from_pat` on it?

As I said in an earlier comment, the problem with `syn` is that it's really verbose and hard to follow (or at least I find it so). So in my view, anything we can do to make this code as structured and clear as possible will be a big help with verifying correctness.

overlookmotel commented:
Running ecosystem CI before merging: https://github.com/oxc-project/oxlint-ecosystem-ci/actions/runs/17525661651

overlookmotel commented:
### Merge activity

* **Sep 7, 7:53 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel commented on crates/oxc_linter/src/lib.rs:
```diff
@@ -170,7 +172,28 @@ impl Linter {
+                let mut rules_any_ast_type = Vec::with_capacity(rules.len());
+
                 for (rule, ctx) in &rules {
+                    // Collect node type information for rules. In large files, benchmarking showed it was worth
+                    // collecting rules into buckets by AST node type to avoid iterating over all rules for each node.
```
#13574

overlookmotel commented on tasks/linter_codegen/src/main.rs:
```diff
@@ -0,0 +1,330 @@
+            let mut result = CollectionResult::Complete;
+            for p in &orpat.cases {
+                if extract_variants_from_pat(p, out) == CollectionResult::Incomplete {
+                    result = CollectionResult::Incomplete;
```
#13576

camchenry committed:
perf(linter): skip rules that do not have any relevant node types (#13138)

- part of https://github.com/oxc-project/oxc/issues/12223
- supersedes stack: https://github.com/oxc-project/oxc/pull/12228

This PR adds the ability to add node type information to indicate what node types a lint rule acts on. The lint rule runner can use this information to optimize running lint rules by skipping nodes, files, or rules that don't apply.

This is a simple version that is focused on creating the framework for automatically generating the rule runner code. This is not the end of optimizations. Currently only 102 rules have node type information, while 471 rules have no node type information and will not be skipped. Adding node type information for more rules will improve performance. We may also be able to skip more nodes and files by being clever about how we use this information. For example, filtering out any rules which do not run on any node type, if the file contains no relevant node types for the rule.

- Adds a new `RuleRunner` trait which includes information needed for indicating whether a rule has node type info or not
- Adds a `oxc_linter_codegen` task which is used to generate the implementation of `RuleRunner` trait for all lint rules. This is mostly written by AI, so worth a bit of skepticism, but I've tried to manually verify the results and it looks good (just lacking more complex detection).
  - Currently, only rules that contain a single top-level `if` statement (or chain of `else if`), or a single `match` statement will have their node type info generated. This should give us a high level of confidence, since these patterns are relatively easy to verify and identify, but it's only a subset of all rules. Next steps will be to add support for `let...else`
- Adjusts the main linting loop to use the node type information to skip rules for node types that don't apply
- Updates CI to include a check for any uncommitted linter codegen changes

---

## Benchmarks

In the latest version of this PR, CodSpeed shows an up to +13% increase in linter performance across large and small files. In the real-world, I've seen a ~5-15% improvement in performance, but in some cases close to zero for very small files. In either case, there appears to be little downside and a lot of potential upside here.

<img width="769" height="242" alt="Screenshot 2025-08-24 at 9 29 30 PM" src="https://github.com/user-attachments/assets/d2c430c4-17a9-4606-85d5-c66c734975d9" />
<img width="757" height="308" alt="Screenshot 2025-08-24 at 9 30 24 PM" src="https://github.com/user-attachments/assets/5508464f-de55-4f4e-9ddd-58f215c4ce85" />
<img width="976" height="368" alt="Screenshot 2025-08-24 at 9 34 52 PM" src="https://github.com/user-attachments/assets/203f83c7-46f3-49f1-a1a8-e18f7ec8483b" />
